feat(ui-tools): add dropDown component

shootermv opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->
![image](https://user-images.githubusercontent.com/1336862/197527657-8831df7b-9406-44b4-ad00-6e0a8ce8e162.png)
Closes  #47356

<!-- Feel free to add any additional description of changes below this line -->

Sboonny commented on tools/ui-components/src/drop-down-button/drop-down-button.tsx:
```diff
@@ -0,0 +1,92 @@
+        </Menu.Button>
+      </div>
+
+      <Transition
+        as={Fragment}
+        enter='transition ease-out duration-100'
+        enterFrom='transform opacity-0 scale-95'
+        enterTo='transform opacity-100 scale-100'
+        leave='transition ease-in duration-75'
+        leaveFrom='transform opacity-100 scale-100'
+        leaveTo='transform opacity-0 scale-95'
+      >
```
Just wondering, do we want to have animation as default?

There are instances like menu carrying `Sign in` and `Sign out`, campers will use it frequently and seeing animation can be obnoxious for some.

And how does tailwind interact with `reduce-motion`, does it alter the `utility classes` value?

ahmadabdolsaheb commented:
@shootermv, Thank you for this pull request. Not sure if this pr is a work in progress, but I wanted to point a few things that are missing. The main aim of the component library is to replace the react-bootstrap library with minimum to no modifications on the lean's client side. 

So our components should be identical to the used react-bootstrap ones:
- they should receive the same props
- have a similar logic
- look the same
- output the same elements

To clarify the technical specifications, it would be great if you could research the current usage as described in the [documentation](https://contribute.freecodecamp.org/#/how-to-work-on-the-component-library?id=how-to-work-on-the-component-library) and create an issue. Then, you could finalize this pr accordingly.

shootermv commented:
can i see some place that project currently has this dropdownButton functionality?

ahmadabdolsaheb commented:
@shootermv, here are multiple use cases: https://github.com/freeCodeCamp/freeCodeCamp/search?q=DropdownButton
Doing a search for `DropdownButton` on your editor will give you better results.

ahmadabdolsaheb commented:
@shootermv, the component is taking shape pretty nicely. Let me know if I could be of any help or if you need any further clarification.

shootermv commented:
currently i added `buttonText` and `options` props, dont sure what to do next

ahmadabdolsaheb commented:
@shootermv so the implementation in this pr has different styles and functionality comparing to the bootstrap component in the codebase. I highly recommend reviewing the https://github.com/freeCodeCamp/freeCodeCamp/issues/44668 issue, the related documentation (mentioned in the same issue) and create an issue accordingly so we could finalize the specifications before proceeding with updating this pr.

ahmadabdolsaheb commented:
@shootermv how is the progress going on your side. Let me know if you need any help.

Sembauke commented:
Hey @shootermv.

Are you still planning to work on this? If not, we can open it to contributors.

Have a nice day and happy coding

shootermv commented:
Hi. Yes. I wnat  to finish this task. Only i currently dont exactly know
what need to be done for finish it

On Thu, Oct 13, 2022, 12:47 Sem Bauke ***@***.***> wrote:

> Hey @shootermv <https://github.com/shootermv>.
>
> Are you still planning to work on this? If not, we can open it to
> contributors.
>
> Have a nice day and happy coding
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/freeCodeCamp/freeCodeCamp/pull/47043#issuecomment-1277342920>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKGMHT3UNOIDYRQCEZIK5TWC7LDVANCNFSM54V6ZCWA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ahmadabdolsaheb commented:
@shootermv, the aim is to just replace the source of the components and have them work. Here are a few considerations:

- The `options` prop is not needed. Each Menu.Item needs to be a stand alone component ( MenuItem ) 
- MenuItem does not need to stay in a separate directory and can be a part of the Drop-Down-Button story.
- The applied animation is beautifully done, but animation is not needed at this point. We will  unify and reimplement all animation on the next phases. 
- bottonText prop should be called Title
- Since we are using headless UI we have less control over the output of the components, so it is ok if they don’t share the same structure with the react-bootstrap ones.
- The dropdown used on the mobile version of legacy certifications (tool-panel.tsx) is actually a drop-up. Feel free to pass a direction prop (‘up’) to the currently used DropDownButton component in the Editor, and clean up the styles there. Once that is done, have the recently developed component accept that prop and default to ‘down’ if direction prop is not passed.
- Currently 5 props are being passed to the old DropDownButton component. Since `block`, ‘bsStyle’, and `className` are being passed the similar values, you don’t need a logic to differentiate if those props are being passed or not. Just make sure Typescript does not error out.
- Free free to expand on the tests by looking at other component tests in our ui library and by searching for `dropdown` in the [react-bootstrap](https://github.com/react-bootstrap/react-bootstrap/tree/master/test) repo and incorporate the relevant ones in our library. 

Once those changes are in, I will give it a final round of review and proceed with the merge.

Let me know if you have other questions.

ahmadabdolsaheb commented:
The currently used components use boostrap styles and some custom styles.
The newly developed components (such as this one) do not need bootstrap or custom styles to work. They should replicated the current look only using tailwind classes.

ahmadabdolsaheb commented:
You can look into similarly built components on the ui-components directory.

ahmadabdolsaheb commented:
> must understand how correctly import this tools/ui-components project to `client` in order to test dropdownbutton there

You need to `npm run build` and `npm run build:css` from the ui-components directory.
Once the builds are finished, import the bundle file and the css file to your desired learn component. 

The path should be something similar to the following. (not sure about the number of ../ s :)
import { MenuItem } from '../../../../../tools/ui-components/dist/bundle.esm';
'../../../../../tools/ui-components/dist/base.css';

If you are getting errors with scss, that is expected. You would add an empty postcss.config.js file in the root repository and things should be smooth.

Let me know if you have further questions.

shootermv commented:
Actually here is one of questions im currently thinking about:
How we plan new coponents to be used at the site?
Will it be simultaneoulsy with bootstrap styles?(which site currently has)
Thanks

On Tue, Nov 15, 2022, 16:31 Ahmad Abdolsaheb ***@***.***>
wrote:

> must understand how correctly import this tools/ui-components project to
> client in order to test dropdownbutton there
>
> You need to npm run build and npm run build:css from the ui-components
> directory.
> Once the builds are finished, import the bundle file and the css file to
> your desired learn component.
>
> The path should be something similar to the following. (not sure about the
> number of ../ s :)
> import { MenuItem } from
> '../../../../../tools/ui-components/dist/bundle.esm';
> '../../../../../tools/ui-components/dist/base.css';
>
> If you are getting errors with scss, that is expected. You would add an
> empty postcss.config.js file in the root repository and things should be
> smooth.
>
> Let me know if you have further questions.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/freeCodeCamp/freeCodeCamp/pull/47043#issuecomment-1315393473>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKGMHREW2YYUJAAF4GDFIDWIONDDANCNFSM54V6ZCWA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ahmadabdolsaheb commented:
No they should be used regardless. We should not be using bootstrap styles.

raisedadead commented:
We are officially on Node.js 18 now. We have changed the merge requirements accordingly.

You will need to rebase this PR against the main branch OR Click the "Update branch" button to get the new checks going.

Thanks for your contribution.

ahmadabdolsaheb commented:
Thanks for the hard work here.
Here are a few nitpicks before we merge this pr:
- Replace `up` with `dropup` so it is consistent with the bootstrap convention. 
- Please check the list for the props currently passed to the MenuItem and DropDownButton in learn, rel seems missing form the new MenuItem.
- Remove the <br/> from the story and adjust the height of the container div to account for the dropdown/dropup
- Since the MenuItem is also added to the component, please include the related story 
- Include applicable tests for both components. see examples: https://github.com/react-bootstrap/react-bootstrap/tree/master/test
- Since the related issue does not include the currently used CSS, please double check that the styles used in the new component corresponds with the old ones.

shootermv commented:
not sure what do you mean here `rel seems missing form the new MenuItem` 
I  looked at react-bootstrap docs and I cannot see `rel` property anywhere. 
will be glad for explanation
thanks

Sboonny commented:
@shootermv, I couldn't create a story even when with importing a parent.

For now, remove the extra CSS properties, and when everything is sorted we can deal with story

> Since the related issue does not include the currently used CSS, please double check that the styles used in the new component corresponds with the old ones.

Sboonny commented:
I just noticed, that I missed up and lost the Co-author in the commit history, while trying to fix the conflict.

I will squash all my commits when I am done and add the deserved attribution, when I am done 👍

Sboonny committed:
Create a drop down menu

Co-authored-by: ahmad abdolsaheb <ahmad.abdolsaheb@gmail.com>
Co-authored-by: moshe vilner <shootermv@gmail.com>

hopefully sort the tests

hopefully sort the test this time

this time? I hope

test compound component

I give up

This reverts commit d88c6783cccf811c3c5dced0ab980adff0172669.

headless doesn't pass onclick to its buttons children

add the logical statment back

revert the removing the ability to click

sort the tests

fix a typo

Sboonny reviewed (approved):
LGTM 🎉, thank you for your contribution to the page! 👍.

Sboonny commented:
I have no idea how to sort the test, I have spent half a day worth of work into making two tests, I could have finished most of the tabs 😭.

> TestingLibraryElementError: Unable to find an element with the text: Hello world. This could be because the text is broken up by multiple elements. In this case, you can provide a function for your text matcher to make your matcher more flexible.

I don't even understand what it wants. 🤣

Sboonny commented:
there is an issue with the navigation in the menu, as you can't use arrow keys to navigate, this should be solved when we merge to react 17, and if it doesn't I will add the code quickly for the keyevents.

Sboonny committed:
sort the tests

ahmadabdolsaheb commented on tools/ui-components/src/drop-down-button/drop-down-button.tsx:
```diff
@@ -0,0 +1,80 @@
+type MenuProps = React.ComponentPropsWithoutRef<typeof Menu>;
+
+export type DropDownButtonProps = MenuProps & {
+  toggleLabel: React.ReactNode;
```
```suggestion
  toggleLabel: React.ReactNode;
  children: React.ReactNode;
```

ahmadabdolsaheb commented:
Thanks for working hard on this component. Here are a few nitpicks:
- Children prop types should be included for the DropDownButton
- The name MenuItems should be replaced with MenuItem. The files names, exports, etc. included.
- DropDownButton and MenuItem should be both included in the index.ts file of the source directory so it is included in the build.

Sboonny committed:
rename to menu item

Sboonny committed:
let dropdown trigger accept children

Sboonny committed:
include dropdown and menu items in the exports

Sboonny committed:
polish the logic around setting the dropup style

Sboonny committed:
add tests

Sboonny committed:
fix: typo DropDown to Dropdown

ahmadabdolsaheb commented:
Instructions for linking:
1) Make sure the main is set up correctly in package.json of the ui-components. `"main": "./src/index.ts",`
2) go to the ui-components and `pnpm link --global'
3) 'pnpm run build:css' and move the base.css file from dist to the related location on learn. 
4) go to the client and `pnpm link @freecodecamp/ui --global`
5) import `{ Dropdown, MenuItem } from '@freecodecamp/ui';` and use it everywhere the Dropdown and DropdownButton is used

Actionable items:
- Make sure the correct props are being used. ( Dropdown.Toggle is missing bsStyle, MenuItem is missing href, etc. )
- Make sure the component is styled similar to learn. ( :focus has reverse styles )
- Make sure the theme switching work properly.
- etc.

Sboonny committed:
sort the styles

Commit:
sort the type and add bsStyle to dropdownToggle

Sboonny committed:
fix: types in button and menu items

Commit:
align text to center

Sboonny commented:
About width, I don't think it's a good idea to set `width`, because the places that use the dropdown, use different width

![image](https://user-images.githubusercontent.com/88248797/232534295-20022474-e246-485a-a1a7-2851a84b169e.png)

![image](https://user-images.githubusercontent.com/88248797/232534621-2d0af3dc-c7e7-4c6a-8ac6-45d4d9520179.png)

The reason that the theme isn't applied, because the styles are using `html.dark, div.dark` and `html.light, div.light`, swapping it to `dark-pallete` and `light-pallete`, or the other way around should solve this.

Sboonny commented on tools/ui-components/src/drop-down/drop-down-button/drop-down-toggle.tsx:
```diff
@@ -0,0 +1,49 @@
+  className,
+  ...rest
+}: DropDownButtonProps): JSX.Element {
+  if (block) defaultClassNames.push('btn-block');
+  if (bsStyle) defaultClassNames.push('border-foreground-primary');
```
Personally, I think this is better deleted, as far as I know which may be wrong, it's always true so instead of making the user ask a question that is always true, we can reduce the load on the use by keeping it adding it to `defaultClassNames`.

```suggestion
  if (block) defaultClassNames.push('btn-block');
  if (bsStyle) defaultClassNames.push('border-foreground-primary');
```

ahmadabdolsaheb replied:
Sure, as long as we are getting the same look as the dropdown in the learning platform.

Sboonny commented:
I haven't changed the extra files, to not make testing it annoying.

ahmadabdolsaheb commented:
> About width, I don't think it's a good idea to set `width`, because the places that use the dropdown, use different width
We could set width to auto and min width to 100% to avoid really small buttons.

Sboonny committed:
remove the bsstyle and block logic from dropdown

Co-authored-by: ahmad abdolsaheb <ahmad.abdolsaheb@gmail.com>

Sboonny committed:
add w-full to sort default button not taking all width

ahmadabdolsaheb commented on tools/ui-components/package.json:
```diff
@@ -27,6 +27,9 @@
   "devDependencies": {
     "@babel/core": "7.21.4",
     "@babel/preset-env": "7.21.4",
+    "@fortawesome/free-solid-svg-icons": "6.4.0",
+    "@fortawesome/react-fontawesome": "0.2.0",
+    "@headlessui/react": "1.7.13",
```
These should not be devDependencies.

Sboonny replied:
Yeab, can we move `@babel/preset-typescript` to dev dependency too?

it's used for config, we don't need to add it as dependecy https://babeljs.io/docs/babel-preset-typescript

ahmadabdolsaheb replied:
Yes that would be great.

Sboonny replied:
done 👍

Sboonny committed:
move the styling dependecy away from dev

Co-authored-by: ahmad abdolsaheb <ahmad.abdolsaheb@gmail.com>

Sembauke reviewed (changes requested):
So this looks good to me. But one bug appeared when reviewing the stories in story book. 

When viewing the drop-up first. Then view the dropdown. The dropdown doesn't dropdown but drops-up instead.

<img width="1459" alt="image" src="https://user-images.githubusercontent.com/46919888/233330088-929dc154-8392-46a6-b167-85d655009952.png">

Also there does not seem to be a story for the "Menu with Action" story.

shootermv commented:
is it good that menuitems are spread at full width of the page?

On Thu, Apr 20, 2023 at 12:54 PM Sem Bauke ***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
>
> So this looks good to me. But one bug appeared when reviewing the stories
> in story book.
>
> When viewing the drop-up first. Then view the dropdown. The dropdown
> doesn't dropdown but drops-up instead.
>
> [image: image]
> <https://user-images.githubusercontent.com/46919888/233330088-929dc154-8392-46a6-b167-85d655009952.png>
>
> Also there does not seem to be a story for the "Menu with Action" story.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/freeCodeCamp/freeCodeCamp/pull/47043#pullrequestreview-1393638453>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKGMHUVOJPTX74GB3VBCCTXCEBU5ANCNFSM54V6ZCWA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
http://gs500coder.blogspot.com

ahmaxed committed:
Merge branch 'main' into feat/ui-tools-add-dropdown-button

Sboonny committed:
stop the mutation

Co-authored-by: sembauke <semboot699@gmail.com>

Sboonny committed:
remove the extra story

Sembauke commented on tools/ui-components/src/drop-down/drop-down.tsx:
```diff
@@ -0,0 +1,43 @@
+
+export type DropdownProps = React.ComponentPropsWithoutRef<typeof Menu>;
+
+const dropDownItems =
```
Can we revert back to an array and use the concatenate method on a separate array for the drop-up?

Something like this:

```js
const dropdownItems = ['value1', 'value2'];
const dropupItems = ['value3'];

const values = [...dropdownItems].concat(dropdownItems);
```

Sboonny replied:
The reason that I was using `push` is to mutate the array, so the users don't need to copy the array one time or in your example two times, to see the component.

Yes, we can, but I don't know if we should.

Sembauke replied:
I think it was mostly for readability, but sure we can resolve this.

Sboonny replied:
True, your suggestion is more readable, and to be honest, I should have put context of why the mutation happens

shootermv commented:
Was very pleased to see your reponse.
Thank you

On Fri, Apr 28, 2023, 16:26 Huyen Nguyen ***@***.***> wrote:

> @shootermv <https://github.com/shootermv> I wasn't able to review the PR,
> but wanted to say thank you for the extreme patience and for the
> contribution to the component library. I really appreciate it 🙂
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/freeCodeCamp/freeCodeCamp/pull/47043#issuecomment-1527562172>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKGMHTHNMWX4ZZV3SC3QVTXDPAPDANCNFSM54V6ZCWA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
